### PR TITLE
Authentication

### PR DESCRIPTION
**Pull request by ptitdam2001**
- Implement Authentication with JWT in backend
- Plug frontend to backend
- Integrate version into saga

### CODE REVIEW

File: business/MetadataDatabase/.vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,34 @@
+{
```
Comment (bat3) [inline business/MetadataDatabase/.vscode/launch.json]: Add to gitignore

File: business/MetadataDatabase/.vscode/tasks.json [hunk 2]
```diff
@@ -0,0 +1,42 @@
+{
```
Comment (bat3) [inline business/MetadataDatabase/.vscode/tasks.json]: Add to gitignore

File: business/MetadataDatabase/Controllers/AuthController.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        }
+
+        [AllowAnonymous]
+        [HttpPost("authenticate")]
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        public ActionResult<UserAuthenticatedDto> Authenticate([FromBody] Authenticate model) {
+            try {
+                var users = this.userService.FindByLogin(model.Username).ToList();
```
Comment (bat3) [inline business/MetadataDatabase/Controllers/AuthController.cs]:
```suggestion
                var userList= this.userService.FindByLogin(model.Username).ToList();
```

File: business/MetadataDatabase/Controllers/AuthController.cs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+                var tokenHandler = new JwtSecurityTokenHandler();
+                var key = Encoding.ASCII.GetBytes(this.settings.Secret);
+                var tokenDescriptor = new SecurityTokenDescriptor {
+                    Subject = new ClaimsIdentity(new Claim[] {
+                        new Claim(ClaimTypes.Name, user.Id.ToString())
+                    }),
+                    Expires = DateTime.UtcNow.AddMinutes(this.settings.TokenValidityDuration),
+                    SigningCredentials = new SigningCredentials(new SymmetricSecurityKey(key), SecurityAlgorithms.HmacSha256Signature)
+                };
+                var token = tokenHandler.CreateToken(tokenDescriptor);
```
Comment (bat3) [inline business/MetadataDatabase/Controllers/AuthController.cs]: Why not use your private method GenerateToken ?

File: business/MetadataDatabase/Data/UserAuthenticatedDto.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+namespace MetadataDatabase.Data
+{
+    public class UserAuthenticatedDto
```
Comment (bat3) [inline business/MetadataDatabase/Data/UserAuthenticatedDto.cs]: For use inheritance for UserAuthenticatedDto and UserRegisterDto
` public class UserAuthenticatedDto : UserDto
    {
        public string token { get; set; }
    }`

File: business/MetadataDatabase/Services/UserServices.cs [hunk 6]
```diff
@@ -96,5 +103,65 @@ public IEnumerable<UserDto> FindByLogin(string login)
         {
             return this.userRepository.GetBySpecification(user => user.login == login).ToDto();
         }
+
+        /// <summary>
+        /// authenticate someone
+        /// </summary>
+        /// <param name="login">user login</param>
+        /// <param name="password">user password</param>
+        public UserDto Authenticate(string login, string password)
```
Comment (bat3) [inline business/MetadataDatabase/Services/UserServices.cs]: Why not use Authenticate class as input ?

File: business/MetadataDatabase/Controllers/UsersController.cs [hunk 7]
```diff
@@ -97,6 +100,7 @@ public ActionResult<UserDto> Post([FromBody] User user)
         /// <param name="id"></param>
         /// <param name="userIn"></param>
         [HttpPut("{id}")]
+        [AllowAnonymous]
```
Comment (bat3) [inline business/MetadataDatabase/Controllers/UsersController.cs]: For me put should not be AllowAnonymous

**Review by bat3 — CHANGES_REQUESTED** [review]

File: business/MetadataDatabase/Controllers/AuthController.cs (hunk 4 above)
Comment (ptitdam2001) [reply business/MetadataDatabase/Controllers/AuthController.cs]: I have cleaned the code

File: business/MetadataDatabase/Configuration/AuthConfiguration.cs [hunk 8]
```diff
@@ -0,0 +1,8 @@
+namespace MetadataDatabase.Models
+{
+    public class AuthConfiguration
+    {
+        public string Secret { get; set; }
```
Comment (cquennouelle) [inline business/MetadataDatabase/Configuration/AuthConfiguration.cs:5]: What is 'Secret' ?

File: business/MetadataDatabase/Controllers/AuthController.cs [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+                firstname = user.firstname,
+                lastname = user.lastname,
+                token = token
+            });
+        }
+
+        [AllowAnonymous]
+        [HttpGet("token/{username}/{password}")]
+        public IActionResult GetToken(string username, string password) {
+            if (username == password) {
```
Comment (cquennouelle) [inline business/MetadataDatabase/Controllers/AuthController.cs]: username == password ?

File: business/MetadataDatabase/Data/UserAuthenticatedDto.cs (hunk 5 above)
Comment (ptitdam2001) [reply business/MetadataDatabase/Data/UserAuthenticatedDto.cs]: You're right

File: client/web/src/App.tsx [hunk 10]
```diff
@@ -1,53 +1,55 @@
 import React from 'react'
```
Comment (cquennouelle) [inline client/web/src/App.tsx:1]: I gave up here. This pull request is too technical for me.

File: business/MetadataDatabase/Services/UserServices.cs (hunk 6 above)
Comment (ptitdam2001) [reply business/MetadataDatabase/Services/UserServices.cs]: yes good

File: business/MetadataDatabase/Controllers/UsersController.cs (hunk 7 above)
Comment (ptitdam2001) [reply business/MetadataDatabase/Controllers/UsersController.cs]: yes I change

File: client/web/src/App.tsx (hunk 10 above)
Comment (ptitdam2001) [reply client/web/src/App.tsx:1]: lol

Comment (ptitdam2001) [reply client/web/src/App.tsx:1]: Here we are importing the react library

File: business/MetadataDatabase/Controllers/AuthController.cs (hunk 9 above)
Comment (ptitdam2001) [reply business/MetadataDatabase/Controllers/AuthController.cs]: I removed it, it is unecessary and bad

File: business/MetadataDatabase/Configuration/AuthConfiguration.cs (hunk 8 above)
Comment (ptitdam2001) [reply business/MetadataDatabase/Configuration/AuthConfiguration.cs:5]: The secret key is combined with the header and the payload to create a unique hash. You are only able to verify this hash if you have the secret key.

For more information : https://stackoverflow.com/questions/31309759/what-is-secret-key-for-jwt-based-authentication-and-how-to-generate-it/31313582

**Review by bat3 — APPROVED** [review]

**Pull request merged**